README.md
# I CEEBI Client App

The official app of the I CEEBI Congress.

> **This project is not longer mantained, to see the progress of this app and an up to date version heade to [HipyCas/ceebi](https://github.com/HipyCas/ceebi). It does still receive minor updates so the latest version published here still gets a bit of upfates.**


CHANGELOG.md
# CEEBI Changelog

This file serves to keep track of changes across versions of the app.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.3.15] - 2022-07-18

### Added

- Translations to Basque
- Show message when not connected to internet and tried to view attendance
- Filter attendance when fetching to ensure that there is not duplicate registries

### Changed

- Improved German translations

### Fixed

- Report more info to Firebase Crashlytics on non-critical errors
- All app strings are translatable
- Improve double colons event title fix
- Restart Android project due to errors when compiling

## [1.3.7] - 2022-07-08

### Added

- Attendance registration and viewing inside a modal
- Dark mode, which can be toggled manually or manually picked from system
- Firebase Crashlytics reporting for uncaught exceptions

### Fixed

- Notch overlapping (added a dynamic padding to the header)
- Title positioning in ID Change page

### Changed

- Styling and other changes to support iOS and dark mode
- Apply wrap for labels in settings and reduce their content length

### Security

- Bump `protobufjs`
- Bump `eventsource`

## [1.1.2] - 2022-06-20

### Added

- Ripple effect when event is clicked in Schedule view
- Modal with event details is shown on click

### Fixed

- Quotation marks properly rendering in one of the events
- Toasts are now placed slightly higher so tab bar can be used while a toast is shown

## [1.0.2] - 2022-04-08

### Fixed

- Logo icons properly show up in notification buttons

## [1.0.0] - 2022-03-16

### Fixed

- Splash screen background color
- Missing translations added
- Remove detail arrows for toggle buttons in settings
- (iOS) Settings page title positioning

## [0.8.8-beta] - 2022-03-14

### Changed

- Date and time when a notification was sent does not show seconds now in the notification detail modal

### Fixed

- Error caused when quickly tapping a notification to open its modal

## [0.8.7-beta] - 2022-03-05

### Added

- Firebase user properties and performance monitoring for specific code
- Small delay to splash screen hide to ensure that the app is completely rendered before the splash screen goes out

### Changed

- Native splash screen is now white (instead of showing a weirdly stretched logo)

### Security

- Bumped `url-parse` from 1.5.3 to 1.5.10 (See [Dependabot alerts](https://github.com/biocienciasgrx/ceebi/security/dependabot?q=is%3Aclosed) for more info)

## [0.8.3-beta] - 2022-03-02

### Added

- Firebase Crashlytics

### Fixed

- App causing lock screen to show up multiple times when app is opened in some Samsung phones.

## [0.8.1-beta] - 2022-02-17

### Added

- App update checker and installer
- Firebase Performance Monitoring

## [0.7.3-alpha] - 2022-02-17

### Added

- Notifications tab: fetches notifications from JSON file in github and displays them in a list, opening a modal for each one when clicked that includes the full notification body and links

### Fixed

- About developer section buttons now properly linking
- Updated and fixed some errors in Galician translations
- Added 2s duration to toast that shows when offline news are loaded due to bad connection as previously it didn't disappear

## [0.6.5-alpha] - 2022-02-17

### Added

- Add app icon on Android
- Complete about page with translations and message about developer

### Changed

- Remove options when news take long to download, instead the app just loads the cached news without letting the user choose
- Improve sliding between dates in schedule so the vertical bar does not seem too long

## [0.6.1-alpha] - 2022-02-06

### Added

- Google Firebase Analytics for Android app
- First basic integration of web analytics for in-app use

## [0.5.1-alpha] - 2022-02-06

### Added

- Fetch [mecenas.json](https://raw.githubusercontent.com/biocienciasgrx/ceebi/master/mecenas.json) from the GitHub repo (@biocienciasgrx/ceebi) and save it locally (in cache) for use offline.

## [0.5.0-alpha] - 2022-02-05

### Added

- Push notifications: receive push notifications when the app is in background (or even in foreground)

## [0.4.4-alpha] - 2022-02-05 [YANKED]

### Added

- Sort locales by name in alphabetical order in locale selection dropdown
- Translations: German

### Changed

- Moved firebase and analytics account ownership to app@biociencias.es. (In a future, you may view the external services and the use that this app makes of your data in PRIVACY.md)

## [0.4.1-alpha] - 2022-02-05

### Fixed

- Android app splash screen set to CEEBI's

## [0.4.0-alpha] - 2022-02-05

### Added

- Icons to events set from the web itself
- Multiple speakers in events
- Provisional ripple effect for events

### Changed

- **BREAKING (Android)**: Changed package id to `es.biociencias.ceebi`

## [0.3.3-alpha] - 2022-02-04 [YANKED]

### Added

- Images for sponsors in about page
- Main structure of about page
- Changelog file
- Added error handling for image download without connection

### Changed

- Reduced characters in notification extract from 50 to 30
- Refactor code to centralize constants
- Clean up: `views/QRCode.vue`, `views/Settings.vue`
- Add 2 and 3 minutes options and remove 90 and 120 minutes options to event reminders time

### Fixed

- Images not properly loading from cache
- Showing _alt_ text in case images don't load

[unreleased]: https://github.com/biocienciasgrx/ceebi/compare/v1.1.2...HEAD
[1.3.15]: https://github.com/biocienciasgrx/ceebi/releases/tag/v1.3.15
[1.3.7]: https://github.com/biocienciasgrx/ceebi/releases/tag/v1.3.7
[1.1.2]: https://github.com/biocienciasgrx/ceebi/releases/tag/v1.1.2
[1.0.2]: https://github.com/biocienciasgrx/ceebi/releases/tag/v1.0.2
[1.0.0]: https://github.com/biocienciasgrx/ceebi/releases/tag/v1.0.0
[0.8.8-beta]: https://github.com/biocienciasgrx/ceebi/releases/tag/v0.8.8-beta

src/types.ts
export type Schedule = Record<DateString, ScheduleDay>;

export type ScheduleDay = Event[];

export interface Event {
  ID: number;
  data: {
    ID: number;
    title: string;
    content: string;
    post: EventPost;
    time: {
      start: TimeString;
      end: TimeString;
      start_raw: TimeString;
      end_raw: TimeString;
      start_timestamp: number;
      end_timestamp: number;
    };
    hourly_schedules: any[];
    tickets: any[];
    color: MayEmptyString;
    permalink: URL;
    thumbnails: EventThumbnails;
    featured_image: EventThumbnails;
    speakers: Record<`${number}`, EventSpeaker>;
    categories: Record<`${number}`, EventCategory>;
    locations: Record<`${number}`, EventLocation>;
    fields: { label: string; value: string }[];
    TZ0: {
      timezone_type: number;
      timezone: `${string}/${string}`;
    };
  };
  dates: {
    start: DateTime;
    end: DateTime;
    allday: "0" | "1";
    hide_time: "0" | "1";
    past: number;
  }[];
  date: {
    start: DateTime;
    end: DateTime;
  };
}

interface DateTime {
  date: string;
  hour: Hour12;
  minutes: Minutes;
  ampm: "am" | "pm";
  timestamp: number;
}

type DateString = string;
// TODO says it's too complex `${0 | 1 | 2}${Digit}${Digit}${Digit}-${0 | 1}${Digit}-${
//   | 0
//   | 1
//   | 2
//   | 3}${Digit}`;

type Hour12 =
  | "00"
  | "01"
  | "02"
  | "03"
  | "04"
  | "05"
  | "06"
  | "07"
  | "08"
  | "09"
  | "10"
  | "11"
  | "12";

type Hour24 =
  | Hour12
  | "13"
  | "14"
  | "15"
  | "16"
  | "17"
  | "18"
  | "19"
  | "20"
  | "21"
  | "22"
  | "23"
  | "24";

type Minutes = `${0 | 1 | 2 | 3 | 4 | 5}${Digit}`;

type Digit = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

type FullDays = `${0 | 1 | 2}${Digit}`;

type TimeString = `${Hour24}:${Minutes}`;

interface EventPost {
  ID: number;
  post_author: Digit;
  post_date: string;
  post_date_gmt: string;
  post_content: string;
  post_title: string;
  post_excerpt: string;
  post_status: "publish" | string;
  comment_status: "closed" | string;
  png_status: "closed" | string;
  post_password: MayEmptyString;
  post_name: string;
  to_ping: MayEmptyString;
  pinged: MayEmptyString;
  post_modified: string;
  post_modified_gmt: string;
  post_content_filtered: MayEmptyString;
  post_parent: number; // Guess its the id of the post
  guid: URL; //URL type?
  menu_order: number;
  post_type: "mec-events" | string;
  post_mime_type: MayEmptyString;
  comment_count: `${number}`;
  filter: "raw" | string;
}

type URL = `${"http" | "https"}://${string}`;

type MayEmptyString = "" | string;

interface EventThumbnails {
  thumbnail: MayEmptyString;
  thumblist: MayEmptyString;
  gridsquare: MayEmptyString;
  meccarouselthumb: MayEmptyString;
  medium: MayEmptyString;
  large: MayEmptyString;
  full: MayEmptyString;
  tileview: MayEmptyString;
}

interface EventSpeaker {
  id: number;
  name: string;
  job_title: MayEmptyString;
  tel: MayEmptyString;
  email: "" | Email;
  facebook: "" | URL;
  twitter: "" | URL;
  gplus: "" | URL;
  thumbnail: "" | URL;
}

type Email = `${string}@${string}`;

interface EventCategory {
  id: number;
  name: string;
  icon: string;
  color: "" | ColorString;
}

type ColorString = `#${string}`;

type HexChar = Digit | "a" | "b" | "c" | "d" | "e" | "f";

interface EventLocation {
  id: number;
  name: string;
  addres: string;
  latitude: `${number}`;
  longitude: `${number}`;
  url: "" | URL;
  thumbnail: MayEmptyString;
}
